{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: jsonapi_client in /home/johannes/.local/lib/python3.10/site-packages (0.9.9)\n",
      "Requirement already satisfied: requests in /usr/lib/python3/dist-packages (from jsonapi_client) (2.25.1)\n",
      "Requirement already satisfied: aiohttp in /home/johannes/.local/lib/python3.10/site-packages (from jsonapi_client) (3.8.5)\n",
      "Requirement already satisfied: jsonschema in /home/johannes/.local/lib/python3.10/site-packages (from jsonapi_client) (4.19.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (6.0.4)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (1.4.0)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (1.9.2)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (23.1.0)\n",
      "Requirement already satisfied: charset-normalizer<4.0,>=2.0 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (3.2.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (1.3.1)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (4.0.3)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in /home/johannes/.local/lib/python3.10/site-packages (from jsonschema->jsonapi_client) (0.10.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /home/johannes/.local/lib/python3.10/site-packages (from jsonschema->jsonapi_client) (2023.7.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in /home/johannes/.local/lib/python3.10/site-packages (from jsonschema->jsonapi_client) (0.30.2)\n",
      "Requirement already satisfied: idna>=2.0 in /usr/lib/python3/dist-packages (from yarl<2.0,>=1.0->aiohttp->jsonapi_client) (3.3)\n"
     ]
    }
   ],
   "source": [
    "!pip install jsonapi_client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations:\n",
    "biom = GALAXY_INPUTS['biom']\n",
    "max_sample_count = GALAXY_INPUTS['count']\n",
    "\n",
    "# rework max_sample_count.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.longitude</th>\n",
       "      <th>attributes.sample-metadata</th>\n",
       "      <th>attributes.latitude</th>\n",
       "      <th>attributes.biosample</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.collection-date</th>\n",
       "      <th>attributes.sample-desc</th>\n",
       "      <th>attributes.environment-biome</th>\n",
       "      <th>attributes.environment-feature</th>\n",
       "      <th>attributes.environment-material</th>\n",
       "      <th>attributes.sample-name</th>\n",
       "      <th>attributes.sample-alias</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>relationships.biome.data.id</th>\n",
       "      <th>relationships.biome.data.type</th>\n",
       "      <th>relationships.studies.data</th>\n",
       "      <th>url</th>\n",
       "      <th>attributes.analysis-completed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>SRS2329447</td>\n",
       "      <td>-151.0007</td>\n",
       "      <td>[{'key': 'geographic location (longitude)', 'v...</td>\n",
       "      <td>-32.4957</td>\n",
       "      <td>SAMN07136875</td>\n",
       "      <td>SRS2329447</td>\n",
       "      <td>2011-06-21</td>\n",
       "      <td>Keywords: GSC:MIxS MIMS:5.0</td>\n",
       "      <td>ocean_biome</td>\n",
       "      <td>ocean</td>\n",
       "      <td>water</td>\n",
       "      <td>S0413</td>\n",
       "      <td>S0413</td>\n",
       "      <td>2023-09-21T16:35:08</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00005294', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS493132</td>\n",
       "      <td>-153.0658</td>\n",
       "      <td>[{'key': 'temperature', 'value': '28.327', 'un...</td>\n",
       "      <td>6.6710</td>\n",
       "      <td>SAMEA2622957</td>\n",
       "      <td>ERS493132</td>\n",
       "      <td>2011-09-10</td>\n",
       "      <td>This sample (TARA_N000002310) was collected du...</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>surface water layer (ENVO:00002042)</td>\n",
       "      <td>particulate matter, including plankton (ENVO:x...</td>\n",
       "      <td>TARA_N000002310</td>\n",
       "      <td>TARA_N000002310</td>\n",
       "      <td>2023-09-21T16:27:38</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00001306', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2018-06-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>SRS2329525</td>\n",
       "      <td>-68.8755</td>\n",
       "      <td>[{'key': 'geographic location (longitude)', 'v...</td>\n",
       "      <td>38.3243</td>\n",
       "      <td>SAMN07136598</td>\n",
       "      <td>SRS2329525</td>\n",
       "      <td>2011-11-12</td>\n",
       "      <td>Keywords: GSC:MIxS MIMS:5.0</td>\n",
       "      <td>ocean_biome</td>\n",
       "      <td>ocean</td>\n",
       "      <td>water</td>\n",
       "      <td>S0122</td>\n",
       "      <td>S0122</td>\n",
       "      <td>2023-09-21T16:25:19</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00005294', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>SRS2329438</td>\n",
       "      <td>5.5272</td>\n",
       "      <td>[{'key': 'geographic location (longitude)', 'v...</td>\n",
       "      <td>-40.0627</td>\n",
       "      <td>SAMN07136934</td>\n",
       "      <td>SRS2329438</td>\n",
       "      <td>2010-10-27</td>\n",
       "      <td>Keywords: GSC:MIxS MIMS:5.0</td>\n",
       "      <td>ocean_biome</td>\n",
       "      <td>ocean</td>\n",
       "      <td>water</td>\n",
       "      <td>S0472</td>\n",
       "      <td>S0472</td>\n",
       "      <td>2023-09-21T16:11:00</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00005294', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS494107</td>\n",
       "      <td>-79.6493</td>\n",
       "      <td>[{'key': 'temperature', 'value': '18.096139', ...</td>\n",
       "      <td>29.8556</td>\n",
       "      <td>SAMEA2623581</td>\n",
       "      <td>ERS494107</td>\n",
       "      <td>2012-01-17</td>\n",
       "      <td>This sample (TARA_N000003149) was collected du...</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>mesopelagic zone (ENVO:00000213)</td>\n",
       "      <td>particulate matter, including plankton (ENVO:x...</td>\n",
       "      <td>TARA_N000003149</td>\n",
       "      <td>TARA_N000003149</td>\n",
       "      <td>2023-09-21T16:09:41</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00001306', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2018-06-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS492023</td>\n",
       "      <td>-85.2732</td>\n",
       "      <td>[{'key': 'temperature', 'value': '19.566255', ...</td>\n",
       "      <td>-5.2669</td>\n",
       "      <td>SAMEA2622230</td>\n",
       "      <td>ERS492023</td>\n",
       "      <td>2011-04-22</td>\n",
       "      <td>This sample (TARA_N000001652) was collected du...</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>deep chlorophyll maximum layer (ENVO:xxxxxxxx)</td>\n",
       "      <td>particulate matter, including plankton (ENVO:x...</td>\n",
       "      <td>TARA_N000001652</td>\n",
       "      <td>TARA_N000001652</td>\n",
       "      <td>2023-09-21T16:04:20</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00001306', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2018-06-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS491719</td>\n",
       "      <td>-111.7202</td>\n",
       "      <td>[{'key': 'temperature', 'value': '25.14875', '...</td>\n",
       "      <td>-25.8051</td>\n",
       "      <td>SAMEA2622010</td>\n",
       "      <td>ERS491719</td>\n",
       "      <td>2011-04-03</td>\n",
       "      <td>This sample (TARA_N000001574) was collected du...</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>surface water layer (ENVO:00002042)</td>\n",
       "      <td>particulate matter, including plankton (ENVO:x...</td>\n",
       "      <td>TARA_N000001574</td>\n",
       "      <td>TARA_N000001574</td>\n",
       "      <td>2023-09-21T16:01:06</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00001306', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2018-06-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>SRS2329521</td>\n",
       "      <td>-69.8629</td>\n",
       "      <td>[{'key': 'geographic location (longitude)', 'v...</td>\n",
       "      <td>39.6923</td>\n",
       "      <td>SAMN07136594</td>\n",
       "      <td>SRS2329521</td>\n",
       "      <td>2011-11-08</td>\n",
       "      <td>Keywords: GSC:MIxS MIMS:5.0</td>\n",
       "      <td>ocean_biome</td>\n",
       "      <td>ocean</td>\n",
       "      <td>water</td>\n",
       "      <td>S0117</td>\n",
       "      <td>S0117</td>\n",
       "      <td>2023-09-21T15:53:38</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00005294', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>SRS2329458</td>\n",
       "      <td>-64.1670</td>\n",
       "      <td>[{'key': 'geographic location (longitude)', 'v...</td>\n",
       "      <td>31.6667</td>\n",
       "      <td>SAMN07136642</td>\n",
       "      <td>SRS2329458</td>\n",
       "      <td>2010-06-13</td>\n",
       "      <td>Keywords: GSC:MIxS MIMS:5.0</td>\n",
       "      <td>ocean_biome</td>\n",
       "      <td>ocean</td>\n",
       "      <td>water</td>\n",
       "      <td>S0167</td>\n",
       "      <td>S0167</td>\n",
       "      <td>2023-09-21T15:44:47</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00005294', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS487919</td>\n",
       "      <td>-6.5669</td>\n",
       "      <td>[{'key': 'temperature', 'value': '20.5251', 'u...</td>\n",
       "      <td>36.5533</td>\n",
       "      <td>SAMEA2619396</td>\n",
       "      <td>ERS487919</td>\n",
       "      <td>2009-09-15</td>\n",
       "      <td>This sample (TARA_X000000323) was collected du...</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>surface water layer (ENVO:00002042)</td>\n",
       "      <td>particulate matter, including plankton (ENVO:x...</td>\n",
       "      <td>TARA_X000000323</td>\n",
       "      <td>TARA_X000000323</td>\n",
       "      <td>2023-09-21T15:35:56</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "      <td>[{'id': 'MGYS00001306', 'type': 'studies'}, {'...</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2018-06-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type          id  attributes.longitude  \\\n",
       "0  samples  SRS2329447             -151.0007   \n",
       "0  samples   ERS493132             -153.0658   \n",
       "0  samples  SRS2329525              -68.8755   \n",
       "0  samples  SRS2329438                5.5272   \n",
       "0  samples   ERS494107              -79.6493   \n",
       "0  samples   ERS492023              -85.2732   \n",
       "0  samples   ERS491719             -111.7202   \n",
       "0  samples  SRS2329521              -69.8629   \n",
       "0  samples  SRS2329458              -64.1670   \n",
       "0  samples   ERS487919               -6.5669   \n",
       "\n",
       "                          attributes.sample-metadata  attributes.latitude  \\\n",
       "0  [{'key': 'geographic location (longitude)', 'v...             -32.4957   \n",
       "0  [{'key': 'temperature', 'value': '28.327', 'un...               6.6710   \n",
       "0  [{'key': 'geographic location (longitude)', 'v...              38.3243   \n",
       "0  [{'key': 'geographic location (longitude)', 'v...             -40.0627   \n",
       "0  [{'key': 'temperature', 'value': '18.096139', ...              29.8556   \n",
       "0  [{'key': 'temperature', 'value': '19.566255', ...              -5.2669   \n",
       "0  [{'key': 'temperature', 'value': '25.14875', '...             -25.8051   \n",
       "0  [{'key': 'geographic location (longitude)', 'v...              39.6923   \n",
       "0  [{'key': 'geographic location (longitude)', 'v...              31.6667   \n",
       "0  [{'key': 'temperature', 'value': '20.5251', 'u...              36.5533   \n",
       "\n",
       "  attributes.biosample attributes.accession attributes.collection-date  \\\n",
       "0         SAMN07136875           SRS2329447                 2011-06-21   \n",
       "0         SAMEA2622957            ERS493132                 2011-09-10   \n",
       "0         SAMN07136598           SRS2329525                 2011-11-12   \n",
       "0         SAMN07136934           SRS2329438                 2010-10-27   \n",
       "0         SAMEA2623581            ERS494107                 2012-01-17   \n",
       "0         SAMEA2622230            ERS492023                 2011-04-22   \n",
       "0         SAMEA2622010            ERS491719                 2011-04-03   \n",
       "0         SAMN07136594           SRS2329521                 2011-11-08   \n",
       "0         SAMN07136642           SRS2329458                 2010-06-13   \n",
       "0         SAMEA2619396            ERS487919                 2009-09-15   \n",
       "\n",
       "                              attributes.sample-desc  \\\n",
       "0                        Keywords: GSC:MIxS MIMS:5.0   \n",
       "0  This sample (TARA_N000002310) was collected du...   \n",
       "0                        Keywords: GSC:MIxS MIMS:5.0   \n",
       "0                        Keywords: GSC:MIxS MIMS:5.0   \n",
       "0  This sample (TARA_N000003149) was collected du...   \n",
       "0  This sample (TARA_N000001652) was collected du...   \n",
       "0  This sample (TARA_N000001574) was collected du...   \n",
       "0                        Keywords: GSC:MIxS MIMS:5.0   \n",
       "0                        Keywords: GSC:MIxS MIMS:5.0   \n",
       "0  This sample (TARA_X000000323) was collected du...   \n",
       "\n",
       "   attributes.environment-biome  \\\n",
       "0                   ocean_biome   \n",
       "0  marine biome (ENVO:00000447)   \n",
       "0                   ocean_biome   \n",
       "0                   ocean_biome   \n",
       "0  marine biome (ENVO:00000447)   \n",
       "0  marine biome (ENVO:00000447)   \n",
       "0  marine biome (ENVO:00000447)   \n",
       "0                   ocean_biome   \n",
       "0                   ocean_biome   \n",
       "0  marine biome (ENVO:00000447)   \n",
       "\n",
       "                   attributes.environment-feature  \\\n",
       "0                                           ocean   \n",
       "0             surface water layer (ENVO:00002042)   \n",
       "0                                           ocean   \n",
       "0                                           ocean   \n",
       "0                mesopelagic zone (ENVO:00000213)   \n",
       "0  deep chlorophyll maximum layer (ENVO:xxxxxxxx)   \n",
       "0             surface water layer (ENVO:00002042)   \n",
       "0                                           ocean   \n",
       "0                                           ocean   \n",
       "0             surface water layer (ENVO:00002042)   \n",
       "\n",
       "                     attributes.environment-material attributes.sample-name  \\\n",
       "0                                              water                  S0413   \n",
       "0  particulate matter, including plankton (ENVO:x...        TARA_N000002310   \n",
       "0                                              water                  S0122   \n",
       "0                                              water                  S0472   \n",
       "0  particulate matter, including plankton (ENVO:x...        TARA_N000003149   \n",
       "0  particulate matter, including plankton (ENVO:x...        TARA_N000001652   \n",
       "0  particulate matter, including plankton (ENVO:x...        TARA_N000001574   \n",
       "0                                              water                  S0117   \n",
       "0                                              water                  S0167   \n",
       "0  particulate matter, including plankton (ENVO:x...        TARA_X000000323   \n",
       "\n",
       "  attributes.sample-alias attributes.last-update  \\\n",
       "0                   S0413    2023-09-21T16:35:08   \n",
       "0         TARA_N000002310    2023-09-21T16:27:38   \n",
       "0                   S0122    2023-09-21T16:25:19   \n",
       "0                   S0472    2023-09-21T16:11:00   \n",
       "0         TARA_N000003149    2023-09-21T16:09:41   \n",
       "0         TARA_N000001652    2023-09-21T16:04:20   \n",
       "0         TARA_N000001574    2023-09-21T16:01:06   \n",
       "0                   S0117    2023-09-21T15:53:38   \n",
       "0                   S0167    2023-09-21T15:44:47   \n",
       "0         TARA_X000000323    2023-09-21T15:35:56   \n",
       "\n",
       "         relationships.biome.data.id relationships.biome.data.type  \\\n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "0  root:Environmental:Aquatic:Marine                        biomes   \n",
       "\n",
       "                          relationships.studies.data  \\\n",
       "0  [{'id': 'MGYS00005294', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00001306', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00005294', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00005294', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00001306', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00001306', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00001306', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00005294', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00005294', 'type': 'studies'}, {'...   \n",
       "0  [{'id': 'MGYS00001306', 'type': 'studies'}, {'...   \n",
       "\n",
       "                                                 url  \\\n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "\n",
       "  attributes.analysis-completed  \n",
       "0                           NaN  \n",
       "0                    2018-06-07  \n",
       "0                           NaN  \n",
       "0                           NaN  \n",
       "0                    2018-06-07  \n",
       "0                    2018-06-07  \n",
       "0                    2018-06-07  \n",
       "0                           NaN  \n",
       "0                           NaN  \n",
       "0                    2018-06-07  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import os\n",
    "\n",
    "###################\n",
    "# Definie Functions\n",
    "###################\n",
    "\n",
    "\n",
    "sample_accessions = []\n",
    "\n",
    "\n",
    "\n",
    "# retuns list of id`s for all Studies to Biom\n",
    "def GetBiomSamplesByIds(biom_id: str):\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        i = 0\n",
    "        biomes_dfs = []\n",
    "        for r in mgnify.iterate(f'biomes/{biom_id}/samples'):\n",
    "            biom_df = pd.json_normalize(r.json)\n",
    "            biom_df['url'] = str(r.links.self)\n",
    "            biomes_dfs.append(biom_df)\n",
    "            i += 1\n",
    "            if(i == max_sample_count):\n",
    "                break\n",
    "        \n",
    "    main_biomes_df = pd.concat(biomes_dfs)\n",
    "\n",
    "    # main_biomes_df = main_biomes_df.dropna()\n",
    "\n",
    "    return main_biomes_df\n",
    "\n",
    "def GetAnalysisFromStudie(stydy_id):\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        i = 0\n",
    "        biomes_dfs = []\n",
    "        for r in mgnify.iterate(f'studies/{stydy_id}/analyses'):\n",
    "            biom_df = pd.json_normalize(r.json)\n",
    "            biom_df['url'] = str(r.links.self)\n",
    "            biomes_dfs.append(biom_df)\n",
    "    if(biomes_dfs == []):\n",
    "        return [None,None]\n",
    "    main_biomes_df = pd.concat(biomes_dfs)\n",
    "\n",
    "    # main_biomes_df = main_biomes_df.dropna()\n",
    "\n",
    "    return [1,main_biomes_df]\n",
    "\n",
    "df_samples = GetBiomSamplesByIds(biom)\n",
    "\n",
    "df_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'id': 'MGYS00005294', 'type': 'studies'}, {'id': 'MGYS00005780', 'type': 'studies'}]\n",
      "statt MGYS00005294\n",
      "get analysis from MGYS00005294\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'id': 'MGYS00001306', 'type': 'studies'}, {'id': 'MGYS00002392', 'type': 'studies'}, {'id': 'MGYS00006491', 'type': 'studies'}]\n",
      "statt MGYS00001306\n",
      "get analysis from MGYS00001306\n",
      "MGYS00001306 has no analysis\n",
      "statt MGYS00002392\n",
      "get analysis from MGYS00002392\n",
      "[{'id': 'MGYS00005294', 'type': 'studies'}, {'id': 'MGYS00005780', 'type': 'studies'}]\n",
      "statt MGYS00005294\n",
      "skipped MGYS00005294 -duplicate- E\n",
      "[{'id': 'MGYS00005294', 'type': 'studies'}, {'id': 'MGYS00005780', 'type': 'studies'}]\n",
      "statt MGYS00005294\n",
      "skipped MGYS00005294 -duplicate- E\n",
      "[{'id': 'MGYS00001306', 'type': 'studies'}, {'id': 'MGYS00002392', 'type': 'studies'}, {'id': 'MGYS00006491', 'type': 'studies'}]\n",
      "statt MGYS00001306\n",
      "skipped MGYS00001306 -in s_w_a\n",
      "statt MGYS00002392\n",
      "skipped MGYS00002392 -duplicate- E\n",
      "[{'id': 'MGYS00001306', 'type': 'studies'}, {'id': 'MGYS00002368', 'type': 'studies'}, {'id': 'MGYS00002392', 'type': 'studies'}, {'id': 'MGYS00006491', 'type': 'studies'}]\n",
      "statt MGYS00001306\n",
      "skipped MGYS00001306 -in s_w_a\n",
      "statt MGYS00002368\n",
      "get analysis from MGYS00002368\n",
      "[{'id': 'MGYS00001306', 'type': 'studies'}, {'id': 'MGYS00002392', 'type': 'studies'}, {'id': 'MGYS00006491', 'type': 'studies'}]\n",
      "statt MGYS00001306\n",
      "skipped MGYS00001306 -in s_w_a\n",
      "statt MGYS00002392\n",
      "skipped MGYS00002392 -duplicate- E\n",
      "[{'id': 'MGYS00005294', 'type': 'studies'}, {'id': 'MGYS00005780', 'type': 'studies'}]\n",
      "statt MGYS00005294\n",
      "skipped MGYS00005294 -duplicate- E\n",
      "[{'id': 'MGYS00005294', 'type': 'studies'}, {'id': 'MGYS00005780', 'type': 'studies'}]\n",
      "statt MGYS00005294\n",
      "skipped MGYS00005294 -duplicate- E\n",
      "[{'id': 'MGYS00001306', 'type': 'studies'}, {'id': 'MGYS00002368', 'type': 'studies'}, {'id': 'MGYS00002392', 'type': 'studies'}, {'id': 'MGYS00006491', 'type': 'studies'}]\n",
      "statt MGYS00001306\n",
      "skipped MGYS00001306 -in s_w_a\n",
      "statt MGYS00002368\n",
      "skipped MGYS00002368 -duplicate- E\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n\\n\\nfor study in studies_Ids:\\n    study_analaysis = GetAnalysisFromStudie(study)\\n    if(study_analaysis == None):\\n        continue\\n    study_analaysises.append(study_analaysis)\\n\\n\\n'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_accessions = (df_samples['attributes.accession'].to_list())\n",
    "\n",
    "studies = df_samples['relationships.studies.data'].to_list()\n",
    "\n",
    "study_analaysises = []\n",
    "\n",
    "studies_without_analysis = []\n",
    "\n",
    "studies_Ids = []\n",
    "\n",
    "for study in studies:\n",
    "    print(study)\n",
    "    # list of studies for sample\n",
    "    s_ids_sample = []\n",
    "    for s in str(study).split(\"'\"):\n",
    "        if('MGYS' in s):\n",
    "            s_ids_sample.append(s) \n",
    "    # for each study\n",
    "    for st in s_ids_sample:\n",
    "        print(f'statt {st}')\n",
    "        # test if we havent checked study\n",
    "        if(st in studies_Ids):\n",
    "            print(f'skipped {st} -duplicate- E')\n",
    "            # we allready have an analysis\n",
    "            break\n",
    "        if(st not in studies_without_analysis):\n",
    "            print(f'get analysis from {st}')\n",
    "            study_analaysis = GetAnalysisFromStudie(st)\n",
    "            if(study_analaysis[0] == None):\n",
    "                # discard if studie has no analysis\n",
    "                studies_without_analysis.append(st)\n",
    "                print(f'{st} has no analysis')\n",
    "                continue\n",
    "            # keep if study has analysis\n",
    "            studies_Ids.append(st)\n",
    "            study_analaysises.append(study_analaysis[1])\n",
    "            break\n",
    "        else:\n",
    "            print(f'skipped {st} -in s_w_a')\n",
    "    #study_analaysises.append(study_analaysis)\n",
    "\n",
    "study_analaysis = pd.concat(study_analaysises)\n",
    "\n",
    "study_analaysis\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "for study in studies_Ids:\n",
    "    study_analaysis = GetAnalysisFromStudie(study)\n",
    "    if(study_analaysis == None):\n",
    "        continue\n",
    "    study_analaysises.append(study_analaysis)\n",
    "\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.experiment-type</th>\n",
       "      <th>attributes.analysis-status</th>\n",
       "      <th>attributes.analysis-summary</th>\n",
       "      <th>attributes.pipeline-version</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.is-private</th>\n",
       "      <th>attributes.complete-time</th>\n",
       "      <th>attributes.instrument-platform</th>\n",
       "      <th>attributes.instrument-model</th>\n",
       "      <th>relationships.run.data.id</th>\n",
       "      <th>relationships.run.data.type</th>\n",
       "      <th>relationships.study.data.id</th>\n",
       "      <th>relationships.study.data.type</th>\n",
       "      <th>relationships.sample.data.id</th>\n",
       "      <th>relationships.sample.data.type</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00452399</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.1</td>\n",
       "      <td>MGYA00452399</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-03-10T11:24:50</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>SRR5788454</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00005294</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS2329767</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00452400</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.1</td>\n",
       "      <td>MGYA00452400</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-02-03T13:07:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>SRR5788154</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00005294</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS2329586</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00452401</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.1</td>\n",
       "      <td>MGYA00452401</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-02-03T13:09:14</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>SRR5788004</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00005294</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS2329737</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00452402</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.1</td>\n",
       "      <td>MGYA00452402</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-02-03T13:11:14</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>SRR5788001</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00005294</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS2329739</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00452403</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.1</td>\n",
       "      <td>MGYA00452403</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-02-03T13:13:23</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>SRR5787999</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00005294</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS2329742</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00163452</td>\n",
       "      <td>metatranscriptomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MGYA00163452</td>\n",
       "      <td>False</td>\n",
       "      <td>2018-05-14T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERR1712096</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00002368</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS490241</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00163453</td>\n",
       "      <td>metatranscriptomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MGYA00163453</td>\n",
       "      <td>False</td>\n",
       "      <td>2018-05-14T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERR1711991</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00002368</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS488908</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00163454</td>\n",
       "      <td>metatranscriptomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MGYA00163454</td>\n",
       "      <td>False</td>\n",
       "      <td>2018-05-14T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERR1712043</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00002368</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS488464</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00163455</td>\n",
       "      <td>metatranscriptomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MGYA00163455</td>\n",
       "      <td>False</td>\n",
       "      <td>2018-05-14T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERR1712021</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00002368</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS490941</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00163456</td>\n",
       "      <td>metatranscriptomic</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MGYA00163456</td>\n",
       "      <td>False</td>\n",
       "      <td>2018-05-14T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERR1711906</td>\n",
       "      <td>runs</td>\n",
       "      <td>MGYS00002368</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS490201</td>\n",
       "      <td>samples</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1937 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             type            id attributes.experiment-type  \\\n",
       "0   analysis-jobs  MGYA00452399                metagenomic   \n",
       "0   analysis-jobs  MGYA00452400                metagenomic   \n",
       "0   analysis-jobs  MGYA00452401                metagenomic   \n",
       "0   analysis-jobs  MGYA00452402                metagenomic   \n",
       "0   analysis-jobs  MGYA00452403                metagenomic   \n",
       "..            ...           ...                        ...   \n",
       "0   analysis-jobs  MGYA00163452         metatranscriptomic   \n",
       "0   analysis-jobs  MGYA00163453         metatranscriptomic   \n",
       "0   analysis-jobs  MGYA00163454         metatranscriptomic   \n",
       "0   analysis-jobs  MGYA00163455         metatranscriptomic   \n",
       "0   analysis-jobs  MGYA00163456         metatranscriptomic   \n",
       "\n",
       "   attributes.analysis-status  \\\n",
       "0                   completed   \n",
       "0                   completed   \n",
       "0                   completed   \n",
       "0                   completed   \n",
       "0                   completed   \n",
       "..                        ...   \n",
       "0                   completed   \n",
       "0                   completed   \n",
       "0                   completed   \n",
       "0                   completed   \n",
       "0                   completed   \n",
       "\n",
       "                          attributes.analysis-summary  \\\n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "..                                                ...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "\n",
       "   attributes.pipeline-version attributes.accession  attributes.is-private  \\\n",
       "0                          4.1         MGYA00452399                  False   \n",
       "0                          4.1         MGYA00452400                  False   \n",
       "0                          4.1         MGYA00452401                  False   \n",
       "0                          4.1         MGYA00452402                  False   \n",
       "0                          4.1         MGYA00452403                  False   \n",
       "..                         ...                  ...                    ...   \n",
       "0                          4.0         MGYA00163452                  False   \n",
       "0                          4.0         MGYA00163453                  False   \n",
       "0                          4.0         MGYA00163454                  False   \n",
       "0                          4.0         MGYA00163455                  False   \n",
       "0                          4.0         MGYA00163456                  False   \n",
       "\n",
       "   attributes.complete-time attributes.instrument-platform  \\\n",
       "0       2020-03-10T11:24:50                       ILLUMINA   \n",
       "0       2020-02-03T13:07:00                       ILLUMINA   \n",
       "0       2020-02-03T13:09:14                       ILLUMINA   \n",
       "0       2020-02-03T13:11:14                       ILLUMINA   \n",
       "0       2020-02-03T13:13:23                       ILLUMINA   \n",
       "..                      ...                            ...   \n",
       "0       2018-05-14T00:00:00                       ILLUMINA   \n",
       "0       2018-05-14T00:00:00                       ILLUMINA   \n",
       "0       2018-05-14T00:00:00                       ILLUMINA   \n",
       "0       2018-05-14T00:00:00                       ILLUMINA   \n",
       "0       2018-05-14T00:00:00                       ILLUMINA   \n",
       "\n",
       "   attributes.instrument-model relationships.run.data.id  \\\n",
       "0                  NextSeq 550                SRR5788454   \n",
       "0                  NextSeq 550                SRR5788154   \n",
       "0                  NextSeq 550                SRR5788004   \n",
       "0                  NextSeq 550                SRR5788001   \n",
       "0                  NextSeq 550                SRR5787999   \n",
       "..                         ...                       ...   \n",
       "0          Illumina HiSeq 2000                ERR1712096   \n",
       "0          Illumina HiSeq 2000                ERR1711991   \n",
       "0          Illumina HiSeq 2000                ERR1712043   \n",
       "0          Illumina HiSeq 2000                ERR1712021   \n",
       "0          Illumina HiSeq 2000                ERR1711906   \n",
       "\n",
       "   relationships.run.data.type relationships.study.data.id  \\\n",
       "0                         runs                MGYS00005294   \n",
       "0                         runs                MGYS00005294   \n",
       "0                         runs                MGYS00005294   \n",
       "0                         runs                MGYS00005294   \n",
       "0                         runs                MGYS00005294   \n",
       "..                         ...                         ...   \n",
       "0                         runs                MGYS00002368   \n",
       "0                         runs                MGYS00002368   \n",
       "0                         runs                MGYS00002368   \n",
       "0                         runs                MGYS00002368   \n",
       "0                         runs                MGYS00002368   \n",
       "\n",
       "   relationships.study.data.type relationships.sample.data.id  \\\n",
       "0                        studies                   SRS2329767   \n",
       "0                        studies                   SRS2329586   \n",
       "0                        studies                   SRS2329737   \n",
       "0                        studies                   SRS2329739   \n",
       "0                        studies                   SRS2329742   \n",
       "..                           ...                          ...   \n",
       "0                        studies                    ERS490241   \n",
       "0                        studies                    ERS488908   \n",
       "0                        studies                    ERS488464   \n",
       "0                        studies                    ERS490941   \n",
       "0                        studies                    ERS490201   \n",
       "\n",
       "   relationships.sample.data.type  \\\n",
       "0                         samples   \n",
       "0                         samples   \n",
       "0                         samples   \n",
       "0                         samples   \n",
       "0                         samples   \n",
       "..                            ...   \n",
       "0                         samples   \n",
       "0                         samples   \n",
       "0                         samples   \n",
       "0                         samples   \n",
       "0                         samples   \n",
       "\n",
       "                                                  url  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "..                                                ...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "\n",
       "[1937 rows x 18 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_analaysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1, ['MGYA00452486']],\n",
       " [2, ['MGYA00167300', 'MGYA00167342']],\n",
       " [1, ['MGYA00452585']],\n",
       " [1, ['MGYA00452480']],\n",
       " [2, ['MGYA00167038', 'MGYA00167501']],\n",
       " [1, ['MGYA00167937']],\n",
       " [1, ['MGYA00167302']],\n",
       " [1, ['MGYA00452594']],\n",
       " [1, ['MGYA00452522']],\n",
       " [1, ['MGYA00167024']]]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis_accesions = []\n",
    "\n",
    "for sample_acc in sample_accessions:\n",
    "    acc = study_analaysis[study_analaysis['relationships.sample.data.id'] == sample_acc]['attributes.accession'].to_list()\n",
    "    analysis_accesions.append([len(acc),acc])\n",
    "\n",
    "analysis_accesions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start grab for MGYA00452486\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00452486/file/SRR5788294_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00167300\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00167300/file/ERR1756161_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00452585\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00452585/file/SRR5788216_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00452480\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00452480/file/SRR5788303_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00167038\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00167038/file/ERR1879702_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00167937\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00167937/file/ERR562553_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00167302\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00167302/file/ERR562500_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00452594\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00452594/file/SRR5788220_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00452522\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00452522/file/SRR5788281_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00167024\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00167024/file/ERR562413_MERGED_FASTQ_SSU_OTU.tsv\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for analysis_sample in analysis_accesions:\n",
    "    url = \"\"\n",
    "    for analysis_id in analysis_sample[1]:\n",
    "        print(f\"start grab for {analysis_id}\")\n",
    "        #########################################\n",
    "        # Get all downloads for one analysis\n",
    "        #########################################\n",
    "\n",
    "        with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "            \n",
    "            dfs = []\n",
    "            for r in mgnify.iterate(f'analyses/{analysis_id}/downloads'):\n",
    "                df = pd.json_normalize(r.json)\n",
    "                df['url'] = str(r.links.self)\n",
    "                dfs.append(df)\n",
    "            \n",
    "        main_df = pd.concat(dfs)\n",
    "\n",
    "        #print(main_df)\n",
    "\n",
    "        ##########################\n",
    "        # get specific data url\n",
    "        #########################\n",
    "\n",
    "        data_type = \"TSV\"\n",
    "\n",
    "        #########################\n",
    "        # get URL for data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "        #########################\n",
    "\n",
    "        data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "        pipelin_version = 5.0\n",
    "        c1 = main_df[\"attributes.file-format.name\"] == data_type\n",
    "        c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "        c3 = main_df[\"relationships.pipeline.data.id\"] == pipelin_version\n",
    "        if(main_df.loc[(c1 & c2), \"url\"].size == 0):\n",
    "            # if we dont get a match try the same with data_label = 'All reads encoding SSU rRNA'\n",
    "            data_label = \"All reads encoding SSU rRNA\"\n",
    "            c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "            if(main_df.loc[(c1 & c2), \"url\"].size == 0):\n",
    "                continue\n",
    "            else:\n",
    "                url = main_df.loc[(c1 & c2), \"url\"].iloc[0]\n",
    "                break\n",
    "        url = main_df.loc[(c1 & c2), \"url\"].iloc[0]\n",
    "        break\n",
    "\n",
    "    if(url == \"\"):\n",
    "        print(f'no analysis found for {analysis_sample}')\n",
    "        continue\n",
    "\n",
    "    ##########################\n",
    "    # download the data\n",
    "    #########################\n",
    "\n",
    "    data_output_folder = 'outputs/collection'\n",
    "    os.makedirs(data_output_folder, exist_ok=True)\n",
    "\n",
    "    response = requests.get(url)\n",
    "\n",
    "    if not response:\n",
    "        print(f\"Could not download file, got response: {response.status_code}\")\n",
    "        break\n",
    "        \n",
    "    print(url)\n",
    "\n",
    "    data_output_path = os.path.join(data_output_folder, f\"{analysis_id}.tsv\")\n",
    "    with open(data_output_path, \"w\") as f:\n",
    "        f.write(response.text)\n",
    "        \n",
    "    # add to galaxy (manually)\n",
    "    #put(data_output_path)\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start grab for MGYA00452486\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.alias</th>\n",
       "      <th>attributes.file-format.name</th>\n",
       "      <th>attributes.file-format.extension</th>\n",
       "      <th>attributes.file-format.compression</th>\n",
       "      <th>attributes.description.label</th>\n",
       "      <th>attributes.description.description</th>\n",
       "      <th>attributes.group-type</th>\n",
       "      <th>attributes.file-checksum.checksum</th>\n",
       "      <th>attributes.file-checksum.checksum-algorithm</th>\n",
       "      <th>relationships.pipeline.data.id</th>\n",
       "      <th>relationships.pipeline.data.type</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_1.tsv.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_1.tsv.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>InterPro matches</td>\n",
       "      <td>InterPro matches (TSV)</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_2.tsv.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_2.tsv.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>InterPro matches</td>\n",
       "      <td>InterPro matches (TSV)</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_3.tsv.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_3.tsv.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>InterPro matches</td>\n",
       "      <td>InterPro matches (TSV)</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_SSU_MAPSeq.mseq.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_SSU_MAPSeq.mseq.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>MAPseq SSU assignments</td>\n",
       "      <td>MAPSeq output file for SSU</td>\n",
       "      <td>Taxonomic analysis SSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_SSU_OTU.tsv</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_SSU_OTU.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Reads encoding SSU rRNA</td>\n",
       "      <td>All reads encoding SSU rRNA</td>\n",
       "      <td>Taxonomic analysis SSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_LSU_MAPSeq.mseq.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_LSU_MAPSeq.mseq.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>MAPseq LSU assignments</td>\n",
       "      <td>MAPSeq output file for LSU</td>\n",
       "      <td>Taxonomic analysis LSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_LSU_OTU.tsv</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_LSU_OTU.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Reads encoding LSU rRNA</td>\n",
       "      <td>All reads encoding LSU rRNA</td>\n",
       "      <td>Taxonomic analysis LSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     type                                          id  \\\n",
       "0  analysis-job-downloads   SRR5788294_MERGED_FASTQ_InterPro_1.tsv.gz   \n",
       "0  analysis-job-downloads   SRR5788294_MERGED_FASTQ_InterPro_2.tsv.gz   \n",
       "0  analysis-job-downloads   SRR5788294_MERGED_FASTQ_InterPro_3.tsv.gz   \n",
       "0  analysis-job-downloads  SRR5788294_MERGED_FASTQ_SSU_MAPSeq.mseq.gz   \n",
       "0  analysis-job-downloads         SRR5788294_MERGED_FASTQ_SSU_OTU.tsv   \n",
       "0  analysis-job-downloads  SRR5788294_MERGED_FASTQ_LSU_MAPSeq.mseq.gz   \n",
       "0  analysis-job-downloads         SRR5788294_MERGED_FASTQ_LSU_OTU.tsv   \n",
       "\n",
       "                             attributes.alias attributes.file-format.name  \\\n",
       "0   SRR5788294_MERGED_FASTQ_InterPro_1.tsv.gz                         TSV   \n",
       "0   SRR5788294_MERGED_FASTQ_InterPro_2.tsv.gz                         TSV   \n",
       "0   SRR5788294_MERGED_FASTQ_InterPro_3.tsv.gz                         TSV   \n",
       "0  SRR5788294_MERGED_FASTQ_SSU_MAPSeq.mseq.gz                         TSV   \n",
       "0         SRR5788294_MERGED_FASTQ_SSU_OTU.tsv                         TSV   \n",
       "0  SRR5788294_MERGED_FASTQ_LSU_MAPSeq.mseq.gz                         TSV   \n",
       "0         SRR5788294_MERGED_FASTQ_LSU_OTU.tsv                         TSV   \n",
       "\n",
       "  attributes.file-format.extension  attributes.file-format.compression  \\\n",
       "0                              tsv                                True   \n",
       "0                              tsv                                True   \n",
       "0                              tsv                                True   \n",
       "0                              tsv                                True   \n",
       "0                              tsv                               False   \n",
       "0                              tsv                                True   \n",
       "0                              tsv                               False   \n",
       "\n",
       "  attributes.description.label attributes.description.description  \\\n",
       "0             InterPro matches             InterPro matches (TSV)   \n",
       "0             InterPro matches             InterPro matches (TSV)   \n",
       "0             InterPro matches             InterPro matches (TSV)   \n",
       "0       MAPseq SSU assignments         MAPSeq output file for SSU   \n",
       "0      Reads encoding SSU rRNA        All reads encoding SSU rRNA   \n",
       "0       MAPseq LSU assignments         MAPSeq output file for LSU   \n",
       "0      Reads encoding LSU rRNA        All reads encoding LSU rRNA   \n",
       "\n",
       "         attributes.group-type attributes.file-checksum.checksum  \\\n",
       "0          Functional analysis                                     \n",
       "0          Functional analysis                                     \n",
       "0          Functional analysis                                     \n",
       "0  Taxonomic analysis SSU rRNA                                     \n",
       "0  Taxonomic analysis SSU rRNA                                     \n",
       "0  Taxonomic analysis LSU rRNA                                     \n",
       "0  Taxonomic analysis LSU rRNA                                     \n",
       "\n",
       "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id  \\\n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "\n",
       "  relationships.pipeline.data.type  \\\n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "\n",
       "                                                 url  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import os\n",
    "\n",
    "# MGYA00678807\n",
    "analysis_sample = ['MGYA00452486']\n",
    "\n",
    "\n",
    "for analysis_id in analysis_sample:\n",
    "    print(f\"start grab for {analysis_id}\")\n",
    "        #########################################\n",
    "        # Get all downloads for one analysis\n",
    "        #########################################\n",
    "\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "            \n",
    "        dfs = []\n",
    "        for r in mgnify.iterate(f'analyses/{analysis_id}/downloads'):\n",
    "            df = pd.json_normalize(r.json)\n",
    "            df['url'] = str(r.links.self)\n",
    "            dfs.append(df)\n",
    "            \n",
    "    main_df = pd.concat(dfs)\n",
    "\n",
    "    #print(main_df)\n",
    "\n",
    "        ##########################\n",
    "        # get specific data url\n",
    "        #########################\n",
    "\n",
    "    data_type = \"TSV\"\n",
    "    data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "    c1 = main_df[\"attributes.file-format.name\"] == data_type\n",
    "    c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "    main_df = main_df.loc[(c1)]\n",
    "main_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
