{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: jsonapi_client in /home/johannes/.local/lib/python3.10/site-packages (0.9.9)\n",
      "Requirement already satisfied: aiohttp in /home/johannes/.local/lib/python3.10/site-packages (from jsonapi_client) (3.8.5)\n",
      "Requirement already satisfied: requests in /usr/lib/python3/dist-packages (from jsonapi_client) (2.25.1)\n",
      "Requirement already satisfied: jsonschema in /home/johannes/.local/lib/python3.10/site-packages (from jsonapi_client) (4.19.0)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (4.0.3)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (23.1.0)\n",
      "Requirement already satisfied: charset-normalizer<4.0,>=2.0 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (3.2.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (1.4.0)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (1.9.2)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (6.0.4)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/johannes/.local/lib/python3.10/site-packages (from aiohttp->jsonapi_client) (1.3.1)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in /home/johannes/.local/lib/python3.10/site-packages (from jsonschema->jsonapi_client) (0.10.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /home/johannes/.local/lib/python3.10/site-packages (from jsonschema->jsonapi_client) (2023.7.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in /home/johannes/.local/lib/python3.10/site-packages (from jsonschema->jsonapi_client) (0.30.2)\n",
      "Requirement already satisfied: idna>=2.0 in /usr/lib/python3/dist-packages (from yarl<2.0,>=1.0->aiohttp->jsonapi_client) (3.3)\n"
     ]
    }
   ],
   "source": [
    "!pip install jsonapi_client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations:\n",
    "biom = GALAXY_INPUTS['biom']\n",
    "max_sample_count = GALAXY_INPUTS['count']\n",
    "\n",
    "# rework max_sample_count.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.sample-metadata</th>\n",
       "      <th>attributes.biosample</th>\n",
       "      <th>attributes.latitude</th>\n",
       "      <th>attributes.longitude</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.collection-date</th>\n",
       "      <th>attributes.sample-desc</th>\n",
       "      <th>attributes.environment-biome</th>\n",
       "      <th>...</th>\n",
       "      <th>attributes.environment-material</th>\n",
       "      <th>attributes.sample-name</th>\n",
       "      <th>attributes.sample-alias</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>relationships.studies.data</th>\n",
       "      <th>relationships.biome.data.id</th>\n",
       "      <th>relationships.biome.data.type</th>\n",
       "      <th>url</th>\n",
       "      <th>attributes.analysis-completed</th>\n",
       "      <th>attributes.geo-loc-name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS693376</td>\n",
       "      <td>[{'key': 'geographic location (latitude)', 'va...</td>\n",
       "      <td>SAMEA3317944</td>\n",
       "      <td>48.7833</td>\n",
       "      <td>9.1833</td>\n",
       "      <td>ERS693376</td>\n",
       "      <td>2013-01-01</td>\n",
       "      <td>Wine Making Soil Sample</td>\n",
       "      <td>soil</td>\n",
       "      <td>...</td>\n",
       "      <td>soil</td>\n",
       "      <td>Wine_Dechlor1</td>\n",
       "      <td>Wine_Dechlor1</td>\n",
       "      <td>2023-09-15T16:34:43</td>\n",
       "      <td>[{'id': 'MGYS00005519', 'type': 'studies'}, {'...</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil:Agricultural</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744877</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446556</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Pasture</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>642</td>\n",
       "      <td>642</td>\n",
       "      <td>2023-08-06T16:36:59</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744878</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446557</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744878</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Pasture</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>644</td>\n",
       "      <td>644</td>\n",
       "      <td>2023-08-06T16:36:13</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744879</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446558</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Pasture</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>646</td>\n",
       "      <td>646</td>\n",
       "      <td>2023-08-06T16:35:29</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744880</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446559</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744880</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Pasture</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>651</td>\n",
       "      <td>651</td>\n",
       "      <td>2023-08-06T16:34:45</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744881</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446560</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744881</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Pasture</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>655</td>\n",
       "      <td>655</td>\n",
       "      <td>2023-08-06T16:34:00</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744882</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446561</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Pasture</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>657</td>\n",
       "      <td>657</td>\n",
       "      <td>2023-08-06T16:33:14</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744883</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446562</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744883</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Pasture</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>661</td>\n",
       "      <td>661</td>\n",
       "      <td>2023-08-06T16:32:29</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744884</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446563</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Pasture</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>663</td>\n",
       "      <td>663</td>\n",
       "      <td>2023-08-06T16:31:41</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samples</td>\n",
       "      <td>ERS744885</td>\n",
       "      <td>[{'key': 'investigation type', 'value': 'metag...</td>\n",
       "      <td>SAMEA3446564</td>\n",
       "      <td>34.8500</td>\n",
       "      <td>148.5500</td>\n",
       "      <td>ERS744885</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glenrock site</td>\n",
       "      <td>Woodlands</td>\n",
       "      <td>...</td>\n",
       "      <td>Soil</td>\n",
       "      <td>668</td>\n",
       "      <td>668</td>\n",
       "      <td>2023-08-06T16:30:58</td>\n",
       "      <td>[{'id': 'MGYS00000498', 'type': 'studies'}]</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>biomes</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>\n",
       "      <td>2015-07-22</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      type         id                         attributes.sample-metadata  \\\n",
       "0  samples  ERS693376  [{'key': 'geographic location (latitude)', 'va...   \n",
       "0  samples  ERS744877  [{'key': 'investigation type', 'value': 'metag...   \n",
       "0  samples  ERS744878  [{'key': 'investigation type', 'value': 'metag...   \n",
       "0  samples  ERS744879  [{'key': 'investigation type', 'value': 'metag...   \n",
       "0  samples  ERS744880  [{'key': 'investigation type', 'value': 'metag...   \n",
       "0  samples  ERS744881  [{'key': 'investigation type', 'value': 'metag...   \n",
       "0  samples  ERS744882  [{'key': 'investigation type', 'value': 'metag...   \n",
       "0  samples  ERS744883  [{'key': 'investigation type', 'value': 'metag...   \n",
       "0  samples  ERS744884  [{'key': 'investigation type', 'value': 'metag...   \n",
       "0  samples  ERS744885  [{'key': 'investigation type', 'value': 'metag...   \n",
       "\n",
       "  attributes.biosample  attributes.latitude  attributes.longitude  \\\n",
       "0         SAMEA3317944              48.7833                9.1833   \n",
       "0         SAMEA3446556              34.8500              148.5500   \n",
       "0         SAMEA3446557              34.8500              148.5500   \n",
       "0         SAMEA3446558              34.8500              148.5500   \n",
       "0         SAMEA3446559              34.8500              148.5500   \n",
       "0         SAMEA3446560              34.8500              148.5500   \n",
       "0         SAMEA3446561              34.8500              148.5500   \n",
       "0         SAMEA3446562              34.8500              148.5500   \n",
       "0         SAMEA3446563              34.8500              148.5500   \n",
       "0         SAMEA3446564              34.8500              148.5500   \n",
       "\n",
       "  attributes.accession attributes.collection-date   attributes.sample-desc  \\\n",
       "0            ERS693376                 2013-01-01  Wine Making Soil Sample   \n",
       "0            ERS744877                        NaN            Glenrock site   \n",
       "0            ERS744878                        NaN            Glenrock site   \n",
       "0            ERS744879                        NaN            Glenrock site   \n",
       "0            ERS744880                        NaN            Glenrock site   \n",
       "0            ERS744881                        NaN            Glenrock site   \n",
       "0            ERS744882                        NaN            Glenrock site   \n",
       "0            ERS744883                        NaN            Glenrock site   \n",
       "0            ERS744884                        NaN            Glenrock site   \n",
       "0            ERS744885                        NaN            Glenrock site   \n",
       "\n",
       "  attributes.environment-biome  ... attributes.environment-material  \\\n",
       "0                         soil  ...                            soil   \n",
       "0                      Pasture  ...                            Soil   \n",
       "0                      Pasture  ...                            Soil   \n",
       "0                      Pasture  ...                            Soil   \n",
       "0                      Pasture  ...                            Soil   \n",
       "0                      Pasture  ...                            Soil   \n",
       "0                      Pasture  ...                            Soil   \n",
       "0                      Pasture  ...                            Soil   \n",
       "0                      Pasture  ...                            Soil   \n",
       "0                    Woodlands  ...                            Soil   \n",
       "\n",
       "  attributes.sample-name attributes.sample-alias attributes.last-update  \\\n",
       "0          Wine_Dechlor1           Wine_Dechlor1    2023-09-15T16:34:43   \n",
       "0                    642                     642    2023-08-06T16:36:59   \n",
       "0                    644                     644    2023-08-06T16:36:13   \n",
       "0                    646                     646    2023-08-06T16:35:29   \n",
       "0                    651                     651    2023-08-06T16:34:45   \n",
       "0                    655                     655    2023-08-06T16:34:00   \n",
       "0                    657                     657    2023-08-06T16:33:14   \n",
       "0                    661                     661    2023-08-06T16:32:29   \n",
       "0                    663                     663    2023-08-06T16:31:41   \n",
       "0                    668                     668    2023-08-06T16:30:58   \n",
       "\n",
       "                          relationships.studies.data  \\\n",
       "0  [{'id': 'MGYS00005519', 'type': 'studies'}, {'...   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "0        [{'id': 'MGYS00000498', 'type': 'studies'}]   \n",
       "\n",
       "                        relationships.biome.data.id  \\\n",
       "0  root:Environmental:Terrestrial:Soil:Agricultural   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "0               root:Environmental:Terrestrial:Soil   \n",
       "\n",
       "  relationships.biome.data.type  \\\n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "0                        biomes   \n",
       "\n",
       "                                                 url  \\\n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/samp...   \n",
       "\n",
       "  attributes.analysis-completed attributes.geo-loc-name  \n",
       "0                           NaN                     NaN  \n",
       "0                    2015-07-22               Australia  \n",
       "0                    2015-07-22               Australia  \n",
       "0                    2015-07-22               Australia  \n",
       "0                    2015-07-22               Australia  \n",
       "0                    2015-07-22               Australia  \n",
       "0                    2015-07-22               Australia  \n",
       "0                    2015-07-22               Australia  \n",
       "0                    2015-07-22               Australia  \n",
       "0                    2015-07-22               Australia  \n",
       "\n",
       "[10 rows x 21 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import os\n",
    "\n",
    "###################\n",
    "# Definie Functions\n",
    "###################\n",
    "\n",
    "\n",
    "sample_accessions = []\n",
    "\n",
    "\n",
    "\n",
    "# retuns list of id`s for all Studies to Biom\n",
    "def GetBiomSamplesByIds(biom_id: str):\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        i = 0\n",
    "        biomes_dfs = []\n",
    "        for r in mgnify.iterate(f'biomes/{biom_id}/samples'):\n",
    "            biom_df = pd.json_normalize(r.json)\n",
    "            biom_df['url'] = str(r.links.self)\n",
    "            biomes_dfs.append(biom_df)\n",
    "            i += 1\n",
    "            if(i == max_sample_count):\n",
    "                break\n",
    "        \n",
    "    main_biomes_df = pd.concat(biomes_dfs)\n",
    "\n",
    "    # main_biomes_df = main_biomes_df.dropna()\n",
    "\n",
    "    return main_biomes_df\n",
    "\n",
    "def GetAnalysisFromStudie(stydy_id):\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        i = 0\n",
    "        biomes_dfs = []\n",
    "        for r in mgnify.iterate(f'studies/{stydy_id}/analyses'):\n",
    "            biom_df = pd.json_normalize(r.json)\n",
    "            biom_df['url'] = str(r.links.self)\n",
    "            biomes_dfs.append(biom_df)\n",
    "    if(biomes_dfs == []):\n",
    "        return [None,None]\n",
    "    main_biomes_df = pd.concat(biomes_dfs)\n",
    "\n",
    "    # main_biomes_df = main_biomes_df.dropna()\n",
    "\n",
    "    return [1,main_biomes_df]\n",
    "\n",
    "df_samples = GetBiomSamplesByIds(biom)\n",
    "\n",
    "df_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'id': 'MGYS00005519', 'type': 'studies'}, {'id': 'MGYS00005520', 'type': 'studies'}]\n",
      "statt MGYS00005519\n",
      "get analysis from MGYS00005519\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "get analysis from MGYS00000498\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "skipped MGYS00000498 -duplicate- E\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "skipped MGYS00000498 -duplicate- E\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "skipped MGYS00000498 -duplicate- E\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "skipped MGYS00000498 -duplicate- E\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "skipped MGYS00000498 -duplicate- E\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "skipped MGYS00000498 -duplicate- E\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "skipped MGYS00000498 -duplicate- E\n",
      "[{'id': 'MGYS00000498', 'type': 'studies'}]\n",
      "statt MGYS00000498\n",
      "skipped MGYS00000498 -duplicate- E\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n\\n\\nfor study in studies_Ids:\\n    study_analaysis = GetAnalysisFromStudie(study)\\n    if(study_analaysis == None):\\n        continue\\n    study_analaysises.append(study_analaysis)\\n\\n\\n'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_accessions = (df_samples['attributes.accession'].to_list())\n",
    "\n",
    "studies = df_samples['relationships.studies.data'].to_list()\n",
    "\n",
    "study_analaysises = []\n",
    "\n",
    "studies_without_analysis = []\n",
    "\n",
    "studies_Ids = []\n",
    "\n",
    "for study in studies:\n",
    "    print(study)\n",
    "    # list of studies for sample\n",
    "    s_ids_sample = []\n",
    "    for s in str(study).split(\"'\"):\n",
    "        if('MGYS' in s):\n",
    "            s_ids_sample.append(s) \n",
    "    # for each study\n",
    "    for st in s_ids_sample:\n",
    "        print(f'statt {st}')\n",
    "        # test if we havent checked study\n",
    "        if(st in studies_Ids):\n",
    "            print(f'skipped {st} -duplicate- E')\n",
    "            # we allready have an analysis\n",
    "            break\n",
    "        if(st not in studies_without_analysis):\n",
    "            print(f'get analysis from {st}')\n",
    "            study_analaysis = GetAnalysisFromStudie(st)\n",
    "            if(study_analaysis[0] == None):\n",
    "                # discard if studie has no analysis\n",
    "                studies_without_analysis.append(st)\n",
    "                print(f'{st} has no analysis')\n",
    "                continue\n",
    "            # keep if study has analysis\n",
    "            studies_Ids.append(st)\n",
    "            study_analaysises.append(study_analaysis[1])\n",
    "            break\n",
    "        else:\n",
    "            print(f'skipped {st} -in s_w_a')\n",
    "    #study_analaysises.append(study_analaysis)\n",
    "\n",
    "study_analaysis = pd.concat(study_analaysises)\n",
    "\n",
    "study_analaysis\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "for study in studies_Ids:\n",
    "    study_analaysis = GetAnalysisFromStudie(study)\n",
    "    if(study_analaysis == None):\n",
    "        continue\n",
    "    study_analaysises.append(study_analaysis)\n",
    "\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.analysis-status</th>\n",
       "      <th>attributes.experiment-type</th>\n",
       "      <th>attributes.pipeline-version</th>\n",
       "      <th>attributes.analysis-summary</th>\n",
       "      <th>attributes.is-private</th>\n",
       "      <th>attributes.complete-time</th>\n",
       "      <th>attributes.instrument-platform</th>\n",
       "      <th>attributes.instrument-model</th>\n",
       "      <th>relationships.study.data.id</th>\n",
       "      <th>relationships.study.data.type</th>\n",
       "      <th>relationships.sample.data.id</th>\n",
       "      <th>relationships.sample.data.type</th>\n",
       "      <th>relationships.assembly.data.id</th>\n",
       "      <th>relationships.assembly.data.type</th>\n",
       "      <th>url</th>\n",
       "      <th>relationships.run.data.id</th>\n",
       "      <th>relationships.run.data.type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00678807</td>\n",
       "      <td>MGYA00678807</td>\n",
       "      <td>completed</td>\n",
       "      <td>assembly</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2023-09-15T16:32:07</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>MGYS00005519</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS693376</td>\n",
       "      <td>samples</td>\n",
       "      <td>ERZ18322102</td>\n",
       "      <td>assemblies</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00678808</td>\n",
       "      <td>MGYA00678808</td>\n",
       "      <td>completed</td>\n",
       "      <td>assembly</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2023-09-15T16:34:44</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>MGYS00005519</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS693376</td>\n",
       "      <td>samples</td>\n",
       "      <td>ERZ1761223</td>\n",
       "      <td>assemblies</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00528214</td>\n",
       "      <td>MGYA00528214</td>\n",
       "      <td>completed</td>\n",
       "      <td>assembly</td>\n",
       "      <td>4.1</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-05-31T09:53:58</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>MGYS00005519</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS693376</td>\n",
       "      <td>samples</td>\n",
       "      <td>ERZ1761223</td>\n",
       "      <td>assemblies</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00651527</td>\n",
       "      <td>MGYA00651527</td>\n",
       "      <td>completed</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2023-08-06T03:36:08</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>MGYS00000498</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS745059</td>\n",
       "      <td>samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>ERR911383</td>\n",
       "      <td>runs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00651534</td>\n",
       "      <td>MGYA00651534</td>\n",
       "      <td>completed</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2023-08-06T03:47:29</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>MGYS00000498</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS745058</td>\n",
       "      <td>samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>ERR911382</td>\n",
       "      <td>runs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00007787</td>\n",
       "      <td>MGYA00007787</td>\n",
       "      <td>completed</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2015-07-22T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>MGYS00000498</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS744990</td>\n",
       "      <td>samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>ERR911314</td>\n",
       "      <td>runs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00007788</td>\n",
       "      <td>MGYA00007788</td>\n",
       "      <td>completed</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2015-07-22T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>MGYS00000498</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS744993</td>\n",
       "      <td>samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>ERR911317</td>\n",
       "      <td>runs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00007789</td>\n",
       "      <td>MGYA00007789</td>\n",
       "      <td>completed</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2015-07-22T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>MGYS00000498</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS744992</td>\n",
       "      <td>samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>ERR911316</td>\n",
       "      <td>runs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00007790</td>\n",
       "      <td>MGYA00007790</td>\n",
       "      <td>completed</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2015-07-22T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>MGYS00000498</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS744970</td>\n",
       "      <td>samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>ERR911294</td>\n",
       "      <td>runs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00007791</td>\n",
       "      <td>MGYA00007791</td>\n",
       "      <td>completed</td>\n",
       "      <td>metagenomic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>False</td>\n",
       "      <td>2015-07-22T00:00:00</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>MGYS00000498</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS745006</td>\n",
       "      <td>samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "      <td>ERR911330</td>\n",
       "      <td>runs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>551 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             type            id attributes.accession  \\\n",
       "0   analysis-jobs  MGYA00678807         MGYA00678807   \n",
       "0   analysis-jobs  MGYA00678808         MGYA00678808   \n",
       "0   analysis-jobs  MGYA00528214         MGYA00528214   \n",
       "0   analysis-jobs  MGYA00651527         MGYA00651527   \n",
       "0   analysis-jobs  MGYA00651534         MGYA00651534   \n",
       "..            ...           ...                  ...   \n",
       "0   analysis-jobs  MGYA00007787         MGYA00007787   \n",
       "0   analysis-jobs  MGYA00007788         MGYA00007788   \n",
       "0   analysis-jobs  MGYA00007789         MGYA00007789   \n",
       "0   analysis-jobs  MGYA00007790         MGYA00007790   \n",
       "0   analysis-jobs  MGYA00007791         MGYA00007791   \n",
       "\n",
       "   attributes.analysis-status attributes.experiment-type  \\\n",
       "0                   completed                   assembly   \n",
       "0                   completed                   assembly   \n",
       "0                   completed                   assembly   \n",
       "0                   completed                metagenomic   \n",
       "0                   completed                metagenomic   \n",
       "..                        ...                        ...   \n",
       "0                   completed                metagenomic   \n",
       "0                   completed                metagenomic   \n",
       "0                   completed                metagenomic   \n",
       "0                   completed                metagenomic   \n",
       "0                   completed                metagenomic   \n",
       "\n",
       "   attributes.pipeline-version  \\\n",
       "0                          5.0   \n",
       "0                          5.0   \n",
       "0                          4.1   \n",
       "0                          5.0   \n",
       "0                          5.0   \n",
       "..                         ...   \n",
       "0                          2.0   \n",
       "0                          2.0   \n",
       "0                          2.0   \n",
       "0                          2.0   \n",
       "0                          2.0   \n",
       "\n",
       "                          attributes.analysis-summary  attributes.is-private  \\\n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "..                                                ...                    ...   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "0   [{'key': 'Submitted nucleotide sequences', 'va...                  False   \n",
       "\n",
       "   attributes.complete-time attributes.instrument-platform  \\\n",
       "0       2023-09-15T16:32:07                       ILLUMINA   \n",
       "0       2023-09-15T16:34:44                       ILLUMINA   \n",
       "0       2020-05-31T09:53:58                       ILLUMINA   \n",
       "0       2023-08-06T03:36:08                       ILLUMINA   \n",
       "0       2023-08-06T03:47:29                       ILLUMINA   \n",
       "..                      ...                            ...   \n",
       "0       2015-07-22T00:00:00                       ILLUMINA   \n",
       "0       2015-07-22T00:00:00                       ILLUMINA   \n",
       "0       2015-07-22T00:00:00                       ILLUMINA   \n",
       "0       2015-07-22T00:00:00                       ILLUMINA   \n",
       "0       2015-07-22T00:00:00                       ILLUMINA   \n",
       "\n",
       "   attributes.instrument-model relationships.study.data.id  \\\n",
       "0               Illumina MiSeq                MGYS00005519   \n",
       "0               Illumina MiSeq                MGYS00005519   \n",
       "0               Illumina MiSeq                MGYS00005519   \n",
       "0          Illumina HiSeq 2500                MGYS00000498   \n",
       "0          Illumina HiSeq 2500                MGYS00000498   \n",
       "..                         ...                         ...   \n",
       "0          Illumina HiSeq 2500                MGYS00000498   \n",
       "0          Illumina HiSeq 2500                MGYS00000498   \n",
       "0          Illumina HiSeq 2500                MGYS00000498   \n",
       "0          Illumina HiSeq 2500                MGYS00000498   \n",
       "0          Illumina HiSeq 2500                MGYS00000498   \n",
       "\n",
       "   relationships.study.data.type relationships.sample.data.id  \\\n",
       "0                        studies                    ERS693376   \n",
       "0                        studies                    ERS693376   \n",
       "0                        studies                    ERS693376   \n",
       "0                        studies                    ERS745059   \n",
       "0                        studies                    ERS745058   \n",
       "..                           ...                          ...   \n",
       "0                        studies                    ERS744990   \n",
       "0                        studies                    ERS744993   \n",
       "0                        studies                    ERS744992   \n",
       "0                        studies                    ERS744970   \n",
       "0                        studies                    ERS745006   \n",
       "\n",
       "   relationships.sample.data.type relationships.assembly.data.id  \\\n",
       "0                         samples                    ERZ18322102   \n",
       "0                         samples                     ERZ1761223   \n",
       "0                         samples                     ERZ1761223   \n",
       "0                         samples                            NaN   \n",
       "0                         samples                            NaN   \n",
       "..                            ...                            ...   \n",
       "0                         samples                            NaN   \n",
       "0                         samples                            NaN   \n",
       "0                         samples                            NaN   \n",
       "0                         samples                            NaN   \n",
       "0                         samples                            NaN   \n",
       "\n",
       "   relationships.assembly.data.type  \\\n",
       "0                        assemblies   \n",
       "0                        assemblies   \n",
       "0                        assemblies   \n",
       "0                               NaN   \n",
       "0                               NaN   \n",
       "..                              ...   \n",
       "0                               NaN   \n",
       "0                               NaN   \n",
       "0                               NaN   \n",
       "0                               NaN   \n",
       "0                               NaN   \n",
       "\n",
       "                                                  url  \\\n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "..                                                ...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "0   https://www.ebi.ac.uk/metagenomics/api/v1/anal...   \n",
       "\n",
       "   relationships.run.data.id relationships.run.data.type  \n",
       "0                        NaN                         NaN  \n",
       "0                        NaN                         NaN  \n",
       "0                        NaN                         NaN  \n",
       "0                  ERR911383                        runs  \n",
       "0                  ERR911382                        runs  \n",
       "..                       ...                         ...  \n",
       "0                  ERR911314                        runs  \n",
       "0                  ERR911317                        runs  \n",
       "0                  ERR911316                        runs  \n",
       "0                  ERR911294                        runs  \n",
       "0                  ERR911330                        runs  \n",
       "\n",
       "[551 rows x 20 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_analaysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[3, ['MGYA00678807', 'MGYA00678808', 'MGYA00528214']],\n",
       " [3, ['MGYA00652169', 'MGYA00143092', 'MGYA00007718']],\n",
       " [3, ['MGYA00652168', 'MGYA00143011', 'MGYA00007717']],\n",
       " [3, ['MGYA00652167', 'MGYA00142989', 'MGYA00007716']],\n",
       " [3, ['MGYA00652166', 'MGYA00142985', 'MGYA00007723']],\n",
       " [3, ['MGYA00652165', 'MGYA00143083', 'MGYA00007722']],\n",
       " [3, ['MGYA00652164', 'MGYA00143076', 'MGYA00007721']],\n",
       " [3, ['MGYA00652163', 'MGYA00142998', 'MGYA00007720']],\n",
       " [3, ['MGYA00652162', 'MGYA00142981', 'MGYA00007773']],\n",
       " [3, ['MGYA00652161', 'MGYA00143017', 'MGYA00007726']]]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis_accesions = []\n",
    "\n",
    "for sample_acc in sample_accessions:\n",
    "    acc = study_analaysis[study_analaysis['relationships.sample.data.id'] == sample_acc]['attributes.accession'].to_list()\n",
    "    analysis_accesions.append([len(acc),acc])\n",
    "\n",
    "analysis_accesions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start grab for MGYA00678807\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00678807/file/ERZ18322102_FASTA_SSU_OTU.tsv\n",
      "start grab for MGYA00652169\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652169/file/ERR911201_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00652168\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652168/file/ERR911202_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00652167\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652167/file/ERR911203_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00652166\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652166/file/ERR911204_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00652165\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652165/file/ERR911205_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00652164\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652164/file/ERR911206_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00652163\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652163/file/ERR911207_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00652162\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652162/file/ERR911208_MERGED_FASTQ_SSU_OTU.tsv\n",
      "start grab for MGYA00652161\n",
      "https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00652161/file/ERR911209_MERGED_FASTQ_SSU_OTU.tsv\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for analysis_sample in analysis_accesions:\n",
    "    url = \"\"\n",
    "    for analysis_id in analysis_sample[1]:\n",
    "        print(f\"start grab for {analysis_id}\")\n",
    "        #########################################\n",
    "        # Get all downloads for one analysis\n",
    "        #########################################\n",
    "\n",
    "        with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "            \n",
    "            dfs = []\n",
    "            for r in mgnify.iterate(f'analyses/{analysis_id}/downloads'):\n",
    "                df = pd.json_normalize(r.json)\n",
    "                df['url'] = str(r.links.self)\n",
    "                dfs.append(df)\n",
    "            \n",
    "        main_df = pd.concat(dfs)\n",
    "\n",
    "        #print(main_df)\n",
    "\n",
    "        ##########################\n",
    "        # get specific data url\n",
    "        #########################\n",
    "\n",
    "        data_type = \"TSV\"\n",
    "        data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "\n",
    "        c1 = main_df[\"attributes.file-format.name\"] == data_type\n",
    "        c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "        if(main_df.loc[(c1 & c2), \"url\"].size == 0):\n",
    "            continue\n",
    "        url = main_df.loc[(c1 & c2), \"url\"].iloc[0]\n",
    "        break\n",
    "\n",
    "    if(url == \"\"):\n",
    "        print(f'no analysis found for {analysis_sample}')\n",
    "        continue\n",
    "\n",
    "    ##########################\n",
    "    # download the data\n",
    "    #########################\n",
    "\n",
    "    data_output_folder = 'outputs/collection'\n",
    "    os.makedirs(data_output_folder, exist_ok=True)\n",
    "\n",
    "    response = requests.get(url)\n",
    "\n",
    "    if not response:\n",
    "        print(f\"Could not download file, got response: {response.status_code}\")\n",
    "        break\n",
    "        \n",
    "    print(url)\n",
    "\n",
    "    data_output_path = os.path.join(data_output_folder, f\"{analysis_id}.tsv\")\n",
    "    with open(data_output_path, \"w\") as f:\n",
    "        f.write(response.text)\n",
    "        \n",
    "    # add to galaxy (manually)\n",
    "    #put(data_output_path)\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start grab for MGYA00452486\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.alias</th>\n",
       "      <th>attributes.file-format.name</th>\n",
       "      <th>attributes.file-format.extension</th>\n",
       "      <th>attributes.file-format.compression</th>\n",
       "      <th>attributes.description.label</th>\n",
       "      <th>attributes.description.description</th>\n",
       "      <th>attributes.group-type</th>\n",
       "      <th>attributes.file-checksum.checksum</th>\n",
       "      <th>attributes.file-checksum.checksum-algorithm</th>\n",
       "      <th>relationships.pipeline.data.id</th>\n",
       "      <th>relationships.pipeline.data.type</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_1.tsv.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_1.tsv.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>InterPro matches</td>\n",
       "      <td>InterPro matches (TSV)</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_2.tsv.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_2.tsv.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>InterPro matches</td>\n",
       "      <td>InterPro matches (TSV)</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_3.tsv.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_InterPro_3.tsv.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>InterPro matches</td>\n",
       "      <td>InterPro matches (TSV)</td>\n",
       "      <td>Functional analysis</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_SSU_MAPSeq.mseq.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_SSU_MAPSeq.mseq.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>MAPseq SSU assignments</td>\n",
       "      <td>MAPSeq output file for SSU</td>\n",
       "      <td>Taxonomic analysis SSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_SSU_OTU.tsv</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_SSU_OTU.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Reads encoding SSU rRNA</td>\n",
       "      <td>All reads encoding SSU rRNA</td>\n",
       "      <td>Taxonomic analysis SSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_LSU_MAPSeq.mseq.gz</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_LSU_MAPSeq.mseq.gz</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>True</td>\n",
       "      <td>MAPseq LSU assignments</td>\n",
       "      <td>MAPSeq output file for LSU</td>\n",
       "      <td>Taxonomic analysis LSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-job-downloads</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_LSU_OTU.tsv</td>\n",
       "      <td>SRR5788294_MERGED_FASTQ_LSU_OTU.tsv</td>\n",
       "      <td>TSV</td>\n",
       "      <td>tsv</td>\n",
       "      <td>False</td>\n",
       "      <td>Reads encoding LSU rRNA</td>\n",
       "      <td>All reads encoding LSU rRNA</td>\n",
       "      <td>Taxonomic analysis LSU rRNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.1</td>\n",
       "      <td>pipelines</td>\n",
       "      <td>https://www.ebi.ac.uk/metagenomics/api/v1/anal...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     type                                          id  \\\n",
       "0  analysis-job-downloads   SRR5788294_MERGED_FASTQ_InterPro_1.tsv.gz   \n",
       "0  analysis-job-downloads   SRR5788294_MERGED_FASTQ_InterPro_2.tsv.gz   \n",
       "0  analysis-job-downloads   SRR5788294_MERGED_FASTQ_InterPro_3.tsv.gz   \n",
       "0  analysis-job-downloads  SRR5788294_MERGED_FASTQ_SSU_MAPSeq.mseq.gz   \n",
       "0  analysis-job-downloads         SRR5788294_MERGED_FASTQ_SSU_OTU.tsv   \n",
       "0  analysis-job-downloads  SRR5788294_MERGED_FASTQ_LSU_MAPSeq.mseq.gz   \n",
       "0  analysis-job-downloads         SRR5788294_MERGED_FASTQ_LSU_OTU.tsv   \n",
       "\n",
       "                             attributes.alias attributes.file-format.name  \\\n",
       "0   SRR5788294_MERGED_FASTQ_InterPro_1.tsv.gz                         TSV   \n",
       "0   SRR5788294_MERGED_FASTQ_InterPro_2.tsv.gz                         TSV   \n",
       "0   SRR5788294_MERGED_FASTQ_InterPro_3.tsv.gz                         TSV   \n",
       "0  SRR5788294_MERGED_FASTQ_SSU_MAPSeq.mseq.gz                         TSV   \n",
       "0         SRR5788294_MERGED_FASTQ_SSU_OTU.tsv                         TSV   \n",
       "0  SRR5788294_MERGED_FASTQ_LSU_MAPSeq.mseq.gz                         TSV   \n",
       "0         SRR5788294_MERGED_FASTQ_LSU_OTU.tsv                         TSV   \n",
       "\n",
       "  attributes.file-format.extension  attributes.file-format.compression  \\\n",
       "0                              tsv                                True   \n",
       "0                              tsv                                True   \n",
       "0                              tsv                                True   \n",
       "0                              tsv                                True   \n",
       "0                              tsv                               False   \n",
       "0                              tsv                                True   \n",
       "0                              tsv                               False   \n",
       "\n",
       "  attributes.description.label attributes.description.description  \\\n",
       "0             InterPro matches             InterPro matches (TSV)   \n",
       "0             InterPro matches             InterPro matches (TSV)   \n",
       "0             InterPro matches             InterPro matches (TSV)   \n",
       "0       MAPseq SSU assignments         MAPSeq output file for SSU   \n",
       "0      Reads encoding SSU rRNA        All reads encoding SSU rRNA   \n",
       "0       MAPseq LSU assignments         MAPSeq output file for LSU   \n",
       "0      Reads encoding LSU rRNA        All reads encoding LSU rRNA   \n",
       "\n",
       "         attributes.group-type attributes.file-checksum.checksum  \\\n",
       "0          Functional analysis                                     \n",
       "0          Functional analysis                                     \n",
       "0          Functional analysis                                     \n",
       "0  Taxonomic analysis SSU rRNA                                     \n",
       "0  Taxonomic analysis SSU rRNA                                     \n",
       "0  Taxonomic analysis LSU rRNA                                     \n",
       "0  Taxonomic analysis LSU rRNA                                     \n",
       "\n",
       "  attributes.file-checksum.checksum-algorithm relationships.pipeline.data.id  \\\n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "0                                                                        4.1   \n",
       "\n",
       "  relationships.pipeline.data.type  \\\n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "0                        pipelines   \n",
       "\n",
       "                                                 url  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  \n",
       "0  https://www.ebi.ac.uk/metagenomics/api/v1/anal...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import os\n",
    "\n",
    "# MGYA00678807\n",
    "analysis_sample = ['MGYA00452486']\n",
    "\n",
    "\n",
    "for analysis_id in analysis_sample:\n",
    "    print(f\"start grab for {analysis_id}\")\n",
    "        #########################################\n",
    "        # Get all downloads for one analysis\n",
    "        #########################################\n",
    "\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "            \n",
    "        dfs = []\n",
    "        for r in mgnify.iterate(f'analyses/{analysis_id}/downloads'):\n",
    "            df = pd.json_normalize(r.json)\n",
    "            df['url'] = str(r.links.self)\n",
    "            dfs.append(df)\n",
    "            \n",
    "    main_df = pd.concat(dfs)\n",
    "\n",
    "    #print(main_df)\n",
    "\n",
    "        ##########################\n",
    "        # get specific data url\n",
    "        #########################\n",
    "\n",
    "    data_type = \"TSV\"\n",
    "    data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "    c1 = main_df[\"attributes.file-format.name\"] == data_type\n",
    "    c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "    main_df = main_df.loc[(c1)]\n",
    "main_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
