{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install jsonapi_client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations:\n",
    "biom = GALAXY_INPUTS['biom']\n",
    "max_sample_count = GALAXY_INPUTS['count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "import requests\n",
    "import os\n",
    "\n",
    "###################\n",
    "# Definie Functions\n",
    "###################\n",
    "\n",
    "########################################################\n",
    "# input: biom string from MGnify\n",
    "# https://www.ebi.ac.uk/metagenomics/browse/biomes/\n",
    "# example: GetBiomSamplesByIds(root:Engineered:Bioreactor)\n",
    "#\n",
    "#\n",
    "# output: df with the first <max_sample_count> \n",
    "########################################################\n",
    "def GetBiomSamplesByIds(biom_id: str):\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        i = 0\n",
    "        biomes_dfs = []\n",
    "        for r in mgnify.iterate(f'biomes/{biom_id}/samples'):\n",
    "            biom_df = pd.json_normalize(r.json)\n",
    "            biom_df['url'] = str(r.links.self)\n",
    "            biomes_dfs.append(biom_df)\n",
    "            i += 1\n",
    "            if(i == max_sample_count):\n",
    "                break\n",
    "    main_biomes_df = pd.concat(biomes_dfs)\n",
    "    return main_biomes_df\n",
    "\n",
    "########################################################\n",
    "# input: study ACCESSION from mgnify\n",
    "# https://www.ebi.ac.uk/metagenomics/browse/studies\n",
    "# example: GetBiomSamplesByIds(MGYS00006539)\n",
    "#\n",
    "#\n",
    "# output: list with 2 objects\n",
    "# when output[0] == None, we have a study without any analysis data\n",
    "# when output[0] == 1,  output[1] = df with all meta-analysis Informations (accession, date, version, ...) for each sample in the study\n",
    "########################################################\n",
    "\n",
    "def GetAnalysisFromStudy(_study_id):\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        biomes_dfs = []\n",
    "        for r in mgnify.iterate(f'studies/{_study_id}/analyses'):\n",
    "            biom_df = pd.json_normalize(r.json)\n",
    "            biom_df['url'] = str(r.links.self)\n",
    "            biomes_dfs.append(biom_df)\n",
    "    # testing if study has no analysis data\n",
    "    if(biomes_dfs == []):\n",
    "        return [None,None]\n",
    "    main_biomes_df = pd.concat(biomes_dfs)\n",
    "\n",
    "    return [1,main_biomes_df]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#################### \n",
    "# we get a DF with the fist <max_sample_count> of the biom <biom>\n",
    "####################\n",
    "\n",
    "df_samples = GetBiomSamplesByIds(biom)\n",
    "\n",
    "df_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#################### \n",
    "# from this df we extract the accessions for all our samples\n",
    "####################\n",
    "\n",
    "sample_accessions = (df_samples['attributes.accession'].to_list())\n",
    "\n",
    "#################### \n",
    "# aswell as a list of all studys in which ours samples got refrenced\n",
    "# for each sample we get a string with the format [{'id':'<Study-Accession-1>','type': 'studies'},{'id':'<Study-Accession-2>','type': 'studies'},...]\n",
    "####################\n",
    "\n",
    "studies = df_samples['relationships.studies.data'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####################\n",
    "# for each of the studies we test if they have analysis data.\n",
    "# if so we save the data in <study_analaysises>,\n",
    "# until we have data for every sample.\n",
    "####################\n",
    "\n",
    "# list of df with analysis data from studys\n",
    "study_analaysises = []\n",
    "# study-accessions from studys without analysis\n",
    "studies_without_analysis = []\n",
    "# study-accessions from studys having analysis data in <study_analaysises>\n",
    "studies_Ids = []\n",
    "\n",
    "for study in studies:\n",
    "    # list of studies for sample\n",
    "    s_ids_sample = []\n",
    "    for s in str(study).split(\"'\"):\n",
    "        if('MGYS' in s):\n",
    "            s_ids_sample.append(s) \n",
    "    # for each study\n",
    "    for st in s_ids_sample:\n",
    "        # test if we havent checked study\n",
    "        if(st in studies_Ids):\n",
    "            # we allready have analysis data for this study\n",
    "            break\n",
    "        if(st not in studies_without_analysis):\n",
    "            # Get analysis data for this study\n",
    "            study_analaysis = GetAnalysisFromStudy(st)\n",
    "            if(study_analaysis[0] == None):\n",
    "                # discard if study has no analysis\n",
    "                studies_without_analysis.append(st)\n",
    "                continue\n",
    "            # keep if study has analysis\n",
    "            studies_Ids.append(st)\n",
    "            study_analaysises.append(study_analaysis[1])\n",
    "            break\n",
    "        \n",
    "# Merge all analysis data into one large df\n",
    "study_analaysis = pd.concat(study_analaysises)\n",
    "study_analaysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################\n",
    "# from <study_analaysis> we extract all analysis-accessions for all samples in <sample_accessions>\n",
    "#\n",
    "# where analysis_accesions[i][0] = number of analysis-accessions for sample i\n",
    "# and analysis_accesions[i][1] = list of analysis-accessions for sample i\n",
    "####################################\n",
    "\n",
    "analysis_accesions = []\n",
    "for sample_acc in sample_accessions:\n",
    "    acc = study_analaysis[study_analaysis['relationships.sample.data.id'] == sample_acc]['attributes.accession'].to_list()\n",
    "    analysis_accesions.append([len(acc),acc])\n",
    "analysis_accesions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "# for each of of our samples we try to get the taxonomic assingments from MGnify\n",
    "# if none of the analysis accessions of a sample return a file, we skip the sample\n",
    "# (this happens when for e.g. the version of the MGnify pipeline is to old)\n",
    "#\n",
    "# Trying for \"OTUs and taxonomic assignments for SSU rRNA\" and \"All reads encoding SSU rRNA\" beacuse of different pipeline versions\n",
    "#\n",
    "# all Analysis files are saved in 'outputs/collection', so that they are automatically exported into Galaxy as one Collection\n",
    "#\n",
    "# analysis files are saved under 'outputs/collection/<analysis-accession>-<sample-accession>.tsv'\n",
    "##################################\n",
    "\n",
    "# for each of our samples\n",
    "for analysis_sample in analysis_accesions:\n",
    "    url = \"\"\n",
    "    # for each analysis-accession for the sample\n",
    "    for analysis_accession in analysis_sample[1]:\n",
    "        print(f\"start grab for {analysis_accession}\")\n",
    "        \n",
    "        # Get all downloads for one analysis\n",
    "        with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "            \n",
    "            dfs = []\n",
    "            for r in mgnify.iterate(f'analyses/{analysis_accession}/downloads'):\n",
    "                df = pd.json_normalize(r.json)\n",
    "                df['url'] = str(r.links.self)\n",
    "                dfs.append(df)\n",
    "            \n",
    "        main_df = pd.concat(dfs)\n",
    "\n",
    "        data_type = \"TSV\"\n",
    "        data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "        \n",
    "        # get URL for data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "        c1 = main_df[\"attributes.file-format.name\"] == data_type\n",
    "        c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "        if(main_df.loc[(c1 & c2), \"url\"].size == 0):\n",
    "            # if we dont get a match try the same with data_label = 'All reads encoding SSU rRNA'\n",
    "            data_label = \"All reads encoding SSU rRNA\"\n",
    "            c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "            if(main_df.loc[(c1 & c2), \"url\"].size == 0):\n",
    "                continue\n",
    "            else:\n",
    "                url = main_df.loc[(c1 & c2), \"url\"].iloc[0]\n",
    "                break\n",
    "        url = main_df.loc[(c1 & c2), \"url\"].iloc[0]\n",
    "        break\n",
    "\n",
    "    if(url == \"\"):\n",
    "        print(f'no analysis found for {analysis_sample}')\n",
    "        continue\n",
    "\n",
    "    ##########################\n",
    "    # download the data\n",
    "    #########################\n",
    "\n",
    "    data_output_folder = 'outputs/collection'\n",
    "    os.makedirs(data_output_folder, exist_ok=True)\n",
    "\n",
    "    response = requests.get(url)\n",
    "\n",
    "    if not response:\n",
    "        print(f\"Could not download file, got response: {response.status_code}\")\n",
    "        break\n",
    "        \n",
    "    print(url)\n",
    "    data_output_path = os.path.join(data_output_folder, f\"{analysis_accession}-{study_analaysis[study_analaysis['attributes.accession'] == analysis_accession]['relationships.sample.data.id'].to_list()[0]}.tsv\")\n",
    "    with open(data_output_path, \"w\") as f:\n",
    "        f.write(response.text)\n",
    "\n",
    "\n",
    "\n",
    "     "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
