{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install jsonapi_client\n",
    "!pip install pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# declarations:\n",
    "study_id = GALAXY_INPUTS['study_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from jsonapi_client import Session\n",
    "import pandas as pd\n",
    "import requests\n",
    "import os\n",
    "\n",
    "###################\n",
    "# Definie Functions\n",
    "###################\n",
    "\n",
    "########################################################\n",
    "# input: study ACCESSION from mgnify\n",
    "# https://www.ebi.ac.uk/metagenomics/browse/studies\n",
    "# example: GetBiomSamplesByIds(MGYS00006539)\n",
    "#\n",
    "#\n",
    "# output: list with 2 objects\n",
    "# when output[0] == None, we have a study without any analysis data\n",
    "# when output[0] == 1,  output[1] = df with all meta-analysis Informations (accession, date, version, ...) for each sample in the study\n",
    "########################################################\n",
    "\n",
    "def GetAnalysisFromStudy(_study_id):\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        biomes_dfs = []\n",
    "        for r in mgnify.iterate(f'studies/{_study_id}/analyses'):\n",
    "            biom_df = pd.json_normalize(r.json)\n",
    "            biom_df['url'] = str(r.links.self)\n",
    "            biomes_dfs.append(biom_df)\n",
    "    # testing if study has no analysis data\n",
    "    if(biomes_dfs == []):\n",
    "        return [None,None]\n",
    "    main_biomes_df = pd.concat(biomes_dfs)\n",
    "\n",
    "    return [1,main_biomes_df]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#################### \n",
    "# we get a DF with the fist <max_sample_count> of the biom <biom>\n",
    "####################\n",
    "\n",
    "df_samples = GetAnalysisFromStudy(study_id)[1]\n",
    "\n",
    "df_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################\n",
    "# from <study_analaysis> we extract all analysis-accessions for all samples in <sample_accessions>\n",
    "#\n",
    "# where analysis_accesions[i][0] = number of analysis-accessions for sample i\n",
    "# and analysis_accesions[i][1] = list of analysis-accessions for sample i\n",
    "####################################\n",
    "\n",
    "analysis_accesions = df_samples['attributes.accession'].to_list()\n",
    "analysis_accesions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "# for each of of our samples we try to get the taxonomic assingments from MGnify\n",
    "# if none of the analysis accessions of a sample return a file, we skip the sample\n",
    "# (this happens when for e.g. the version of the MGnify pipeline is to old)\n",
    "#\n",
    "# Trying for \"OTUs and taxonomic assignments for SSU rRNA\" and \"All reads encoding SSU rRNA\" beacuse of different pipeline versions\n",
    "#\n",
    "# all Analysis files are saved in 'outputs/collection', so that they are automatically exported into Galaxy as one Collection\n",
    "#\n",
    "# analysis files are saved under 'outputs/collection/<analysis-accession>-<sample-accession>.tsv'\n",
    "##################################\n",
    "\n",
    "\n",
    "# for each of our samples\n",
    "for analysis_accession in analysis_accesions:\n",
    "    url = \"\"\n",
    "    print(f\"start grab for {analysis_accession}\")\n",
    "    # \n",
    "    # Get all downloads for one analysis\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:    \n",
    "        dfs = []\n",
    "        for r in mgnify.iterate(f'analyses/{analysis_accession}/downloads'):\n",
    "            df = pd.json_normalize(r.json)\n",
    "            df['url'] = str(r.links.self)\n",
    "            dfs.append(df)\n",
    "            \n",
    "    main_df = pd.concat(dfs)\n",
    "    data_type = \"TSV\"\n",
    "    data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "        \n",
    "    # get URL for data_label = \"OTUs and taxonomic assignments for SSU rRNA\"\n",
    "    c1 = main_df[\"attributes.file-format.name\"] == data_type\n",
    "    c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "    if(main_df.loc[(c1 & c2), \"url\"].size == 0):\n",
    "        # if we dont get a match try the same with data_label = 'All reads encoding SSU rRNA'\n",
    "        data_label = \"All reads encoding SSU rRNA\"\n",
    "        c2 = main_df[\"attributes.description.description\"] == data_label\n",
    "        if(main_df.loc[(c1 & c2), \"url\"].size == 0):\n",
    "            print(\"No analysis found\")\n",
    "        else:\n",
    "            url = main_df.loc[(c1 & c2), \"url\"].iloc[0]\n",
    "    url = main_df.loc[(c1 & c2), \"url\"].iloc[0]\n",
    "\n",
    "    if(url == \"\"):\n",
    "        print(f'no analysis found for {analysis_accession}')\n",
    "        continue\n",
    "\n",
    "    ##########################\n",
    "    # download the data\n",
    "    #########################\n",
    "\n",
    "    data_output_folder = 'outputs/collection'\n",
    "    os.makedirs(data_output_folder, exist_ok=True)\n",
    "\n",
    "    response = requests.get(url)\n",
    "\n",
    "    if not response:\n",
    "        print(f\"Could not download file, got response: {response.status_code}\")\n",
    "        break\n",
    "        \n",
    "    print(url)\n",
    "    data_output_path = os.path.join(data_output_folder, f\"{analysis_accession}-{df_samples[df_samples['attributes.accession'] == analysis_accession]['relationships.sample.data.id'].to_list()[0]}.tsv\")\n",
    "    with open(data_output_path, \"w\") as f:\n",
    "        f.write(response.text)\n",
    "\n",
    "     "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
